File: engine/benchmark.py
"""Buy-and-hold BTC benchmark construction.

Pure computation layer:
  - BenchmarkSeries: frozen dataclass holding both aligned equity curves
  - build_btc_benchmark: pure function; no I/O

DB helpers (thin, read-only, accept a Connection for testability):
  - load_equity_snapshots: reads equity_curve table
  - load_btc_closes: reads prices table for BTC/USDT
"""

from __future__ import annotations

import bisect
import sqlite3
from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class BenchmarkSeries:
    """Aligned equity curves for soros vs buy-and-hold BTC.

    Both series share the same initial capital and the same window start.
    If BTC data starts after the first equity snapshot, the common window
    begins at the first snapshot that has BTC coverage.
    """

    timestamps: tuple[int, ...]       # unix seconds, ascending
    soros_equity: tuple[float, ...]   # equity_curve values at each timestamp
    btc_equity: tuple[float, ...]     # buy-and-hold BTC equity, same initial capital
    initial_capital: float            # == soros_equity[0]; starting value for both curves
    btc_start_price: float            # BTC/USDT close used as the buy-and-hold reference
    window_start: int                 # == timestamps[0]
    window_end: int                   # == timestamps[-1]
    n_points: int                     # == len(timestamps)
    n_btc_gaps: int                   # snapshots where BTC price was forward-filled


def build_btc_benchmark(
    snapshots: Sequence[tuple[int, float]],
    btc_closes: Sequence[tuple[int, float]],
) -> BenchmarkSeries:
    """Build a buy-and-hold BTC equity curve aligned to soros equity snapshots.

    For each equity snapshot timestamp, locates the most recent BTC/USDT close
    at or before that timestamp (forward-fill for gaps).  Snapshots that fall
    before the first available BTC price are skipped and excluded from the
    aligned window.

    Args:
        snapshots: (unix_ts, equity) pairs from equity_curve, any order.
        btc_closes: (unix_ts, close) pairs for BTC/USDT, any order.

    Returns:
        BenchmarkSeries with both curves normalised to the same initial capital
        at the common window start.

    Raises:
        ValueError: if snapshots is empty, or no BTC close overlaps the window.
    """
    if not snapshots:
        raise ValueError("snapshots is empty")

    snap = sorted(snapshots, key=lambda x: x[0])

    # Build sorted parallel arrays for BTC — bisect needs a list of keys
    sorted_btc = sorted(btc_closes, key=lambda x: x[0])
    btc_ts_arr: list[int] = [ts for ts, _ in sorted_btc]
    btc_price_arr: list[float] = [price for _, price in sorted_btc]

    out_timestamps: list[int] = []
    out_soros: list[float] = []
    out_btc_raw: list[float] = []
    n_btc_gaps = 0
    btc_start_price: float | None = None

    for snap_ts, snap_eq in snap:
        idx = bisect.bisect_right(btc_ts_arr, snap_ts) - 1
        if idx < 0:
            # No BTC price at or before this snapshot — outside common window
            continue

        btc_price = btc_price_arr[idx]
        if btc_ts_arr[idx] != snap_ts:
            n_btc_gaps += 1

        if btc_start_price is None:
            btc_start_price = btc_price

        out_timestamps.append(snap_ts)
        out_soros.append(snap_eq)
        out_btc_raw.append(btc_price)

    if btc_start_price is None:
        raise ValueError("no BTC close prices overlap with the equity snapshot window")

    initial_capital = out_soros[0]
    btc_eq = tuple(initial_capital * (p / btc_start_price) for p in out_btc_raw)

    return BenchmarkSeries(
        timestamps=tuple(out_timestamps),
        soros_equity=tuple(out_soros),
        btc_equity=btc_eq,
        initial_capital=initial_capital,
        btc_start_price=btc_start_price,
        window_start=out_timestamps[0],
        window_end=out_timestamps[-1],
        n_points=len(out_timestamps),
        n_btc_gaps=n_btc_gaps,
    )


# ---------------------------------------------------------------------------
# DB helpers — thin read-only wrappers; pure functions call build_btc_benchmark
# ---------------------------------------------------------------------------

def load_equity_snapshots(
    conn: sqlite3.Connection,
    *,
    is_paper: bool | None = None,
) -> list[tuple[int, float]]:
    """Return (ts, equity) rows from equity_curve, ordered by ts ascending."""
    if is_paper is None:
        rows = conn.execute(
            "SELECT ts, equity FROM equity_curve ORDER BY ts ASC, id ASC"
        ).fetchall()
    else:
        rows = conn.execute(
            "SELECT ts, equity FROM equity_curve"
            " WHERE is_paper = ? ORDER BY ts ASC, id ASC",
            (int(is_paper),),
        ).fetchall()
    return [(int(row[0]), float(row[1])) for row in rows]


def load_btc_closes(
    conn: sqlite3.Connection,
    *,
    symbol: str = "BTC/USDT",
    timeframe: str = "1h",
    since: int | None = None,
    until: int | None = None,
) -> list[tuple[int, float]]:
    """Return (ts, close) rows for BTC/USDT from prices, ordered by ts ascending."""
    params: list = [symbol, timeframe]
    where = "symbol = ? AND timeframe = ?"
    if since is not None:
        where += " AND ts >= ?"
        params.append(since)
    if until is not None:
        where += " AND ts <= ?"
        params.append(until)
    rows = conn.execute(
        f"SELECT ts, close FROM prices WHERE {where} ORDER BY ts ASC",  # noqa: S608
        params,
    ).fetchall()
    return [(int(row[0]), float(row[1])) for row in rows]

File: dashboard/lib/benchmark.ts
/**
 * Pure benchmark and metrics computation — no I/O, no Next.js dependencies.
 * Mirrors engine/benchmark.py and engine/metrics.py.
 */

// ---------------------------------------------------------------------------
// Benchmark construction
// ---------------------------------------------------------------------------

export interface BenchmarkSeries {
  timestamps: number[];
  sorosEquity: number[];
  btcEquity: number[];
  initialCapital: number;
  btcStartPrice: number;
  windowStart: number;
  windowEnd: number;
  nPoints: number;
  nBtcGaps: number;
}

/**
 * Build a buy-and-hold BTC equity curve aligned to soros equity snapshots.
 *
 * For each snapshot timestamp, locates the most recent BTC/USDT close at or
 * before that timestamp (forward-fill for gaps).  Snapshots that fall before
 * the first available BTC price are excluded from the aligned window.
 */
export function buildBtcBenchmark(
  snapshots: [number, number][],
  btcCloses: [number, number][]
): BenchmarkSeries {
  if (snapshots.length === 0) throw new Error("snapshots is empty");

  const snap = [...snapshots].sort((a, b) => a[0] - b[0]);
  const sortedBtc = [...btcCloses].sort((a, b) => a[0] - b[0]);
  const btcTsArr = sortedBtc.map((x) => x[0]);
  const btcPriceArr = sortedBtc.map((x) => x[1]);

  const outTimestamps: number[] = [];
  const outSoros: number[] = [];
  const outBtcRaw: number[] = [];
  let nBtcGaps = 0;
  let btcStartPrice: number | null = null;

  for (const [snapTs, snapEq] of snap) {
    // bisect_right: rightmost idx where btcTsArr[idx] <= snapTs
    let lo = 0;
    let hi = btcTsArr.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (btcTsArr[mid] <= snapTs) lo = mid + 1;
      else hi = mid;
    }
    const idx = lo - 1;
    if (idx < 0) continue; // no BTC price at or before this snapshot

    const btcPrice = btcPriceArr[idx];
    if (btcTsArr[idx] !== snapTs) nBtcGaps++;

    if (btcStartPrice === null) btcStartPrice = btcPrice;

    outTimestamps.push(snapTs);
    outSoros.push(snapEq);
    outBtcRaw.push(btcPrice);
  }

  if (btcStartPrice === null)
    throw new Error("no BTC close prices overlap with the equity snapshot window");

  const initialCapital = outSoros[0];
  const btcEquity = outBtcRaw.map((p) => initialCapital * (p / btcStartPrice!));

  return {
    timestamps: outTimestamps,
    sorosEquity: outSoros,
    btcEquity,
    initialCapital,
    btcStartPrice,
    windowStart: outTimestamps[0],
    windowEnd: outTimestamps[outTimestamps.length - 1],
    nPoints: outTimestamps.length,
    nBtcGaps,
  };
}

// ---------------------------------------------------------------------------
// Metrics
// ---------------------------------------------------------------------------

export const MIN_SHARPE_N = 30;
const SECONDS_PER_YEAR = 365.25 * 24 * 3600;

export interface ComparisonMetrics {
  sorosTotalReturn: number;
  sorosSharpe: number | null;
  sorosMaxDrawdown: number;
  btcTotalReturn: number;
  btcSharpe: number | null;
  btcMaxDrawdown: number;
  n: number;
  annualizationFactor: number;
  medianIntervalSeconds: number;
  sharpeConclusive: boolean;
  riskFreeRate: number;
}

function totalReturn(equity: number[]): number {
  if (equity.length < 1) return 0;
  return (equity[equity.length - 1] - equity[0]) / equity[0];
}

function maxDrawdown(equity: number[]): number {
  if (equity.length < 2) return 0;
  let peak = equity[0];
  let maxDd = 0;
  for (const v of equity) {
    if (v > peak) peak = v;
    const dd = (v - peak) / peak;
    if (dd < maxDd) maxDd = dd;
  }
  return maxDd;
}

function periodReturns(equity: number[]): number[] {
  const out: number[] = [];
  for (let i = 1; i < equity.length; i++) out.push((equity[i] - equity[i - 1]) / equity[i - 1]);
  return out;
}

function mean(arr: number[]): number {
  return arr.reduce((s, v) => s + v, 0) / arr.length;
}

function stdev(arr: number[]): number {
  if (arr.length < 2) return 0;
  const m = mean(arr);
  return Math.sqrt(arr.reduce((s, v) => s + (v - m) ** 2, 0) / (arr.length - 1));
}

function sharpe(equity: number[], annFactor: number): number | null {
  if (equity.length < 2) return null;
  const rets = periodReturns(equity);
  if (rets.length < 1) return null;
  const std = stdev(rets);
  if (std === 0) return null;
  return (mean(rets) / std) * Math.sqrt(annFactor);
}

function medianInterval(timestamps: number[]): number {
  if (timestamps.length < 2) return 3600;
  const diffs: number[] = [];
  for (let i = 1; i < timestamps.length; i++) diffs.push(timestamps[i] - timestamps[i - 1]);
  diffs.sort((a, b) => a - b);
  const mid = Math.floor(diffs.length / 2);
  return diffs.length % 2 === 0 ? (diffs[mid - 1] + diffs[mid]) / 2 : diffs[mid];
}

function annualizationFactor(medianIntervalSec: number): number {
  if (medianIntervalSec <= 0) return SECONDS_PER_YEAR / 3600;
  return SECONDS_PER_YEAR / medianIntervalSec;
}

export function computeMetrics(series: BenchmarkSeries): ComparisonMetrics {
  const medInterval = medianInterval(series.timestamps);
  const annFactor = annualizationFactor(medInterval);

  return {
    sorosTotalReturn: totalReturn(series.sorosEquity),
    sorosSharpe: sharpe(series.sorosEquity, annFactor),
    sorosMaxDrawdown: maxDrawdown(series.sorosEquity),
    btcTotalReturn: totalReturn(series.btcEquity),
    btcSharpe: sharpe(series.btcEquity, annFactor),
    btcMaxDrawdown: maxDrawdown(series.btcEquity),
    n: series.nPoints,
    annualizationFactor: annFactor,
    medianIntervalSeconds: medInterval,
    sharpeConclusive: series.nPoints >= MIN_SHARPE_N,
    riskFreeRate: 0.0,
  };
}
